README.md
![](https://raw.githubusercontent.com/dejanfajfar/epoch.net/master/images/logo.png)

---

![GitHub Workflow Status (with event)](https://img.shields.io/github/actions/workflow/status/dejanfajfar/epoch.net/dotnet.yml?style=flat-square)
![Nuget](https://img.shields.io/nuget/dt/epoch.net?style=flat-square&link=https%3A%2F%2Fwww.nuget.org%2Fpackages%2FEpoch.net)
![Nuget (with prereleases)](https://img.shields.io/nuget/vpre/epoch.net?style=flat-square&link=https%3A%2F%2Fwww.nuget.org%2Fpackages%2FEpoch.net)

---

# 👋 Introduction

> A simple and non-intrusive library for all your epoch needs

```csharp
using Epoch.net;

...
var timestamp = EpochTime.Now;
...

```

## 🤔 What is an Epoch
Epoch is shorthand for _Unix epoch time_ or as it is also known __POSIX time__.

__Short version:__ The number of seconds since __1970-01-01T00:00:00Z__ without leap seconds.

__Long version:__ [unix time](https://en.wikipedia.org/wiki/Unix_time)

## Motivation
With __.net core__ working on all major operating systems and most of them using POSIX to denote time and timestamps.
Not to mention the myriad of 3rd party tools, using POSIX time internally. The need for a working and tested implementation became clearer and clearer.
Additionally I did not want to write this logic in an inferior form in many of my applications.
So this little library was written.

Its sole purpose is to make the work with these unix timestamps as easy as possible.

For this purpose this library contains extension methods that build upon the existing __DateTime__, __TimeSpan__, __int__ and __long__ types.

# 📖 Glossary

| Term | Description |
|:-----|:-------------|
| EpochTime | Represents a POSIX time instance exact to the _second_ |
| EpochTimestamp | The number of seconds since 1970-01-01T00:00Z |
| LongEpochTime | Represents a POSIX time instance exact to the _millisecond_ |
| LongEpochTimestamp | The number of milliseconds since 1970-01-01T00:00Z |

## Difference between `EpochTime` and `LongEpochTime`

The initial difference was precision. _EpochTime_ will only be accurate to the nearest second on the other hand _LongEpochTime_ is accurate to the millisecond.
From this precision difference I decided to differentiate then by their __underlying__ base type.

`EpochTime` is based on `int32` / `int`

`LongEpochTime` is based on `int64` / `long`  

> ❗This is sadly implies that if you get a POSIX timestamp from somewhere you, the developer, have to know if it is representing milliseconds or seconds!

## Ranges

| Unit | Min value | Max value |
|:----|----:|----:|
| EpochTimestamp | -2147483648 | 2147483647 |
| LongEpochTimestamp | -922337203685477 | 922337203685477 |

📝 _Note:_ The range of the `LongEpoch` is theoretically limited only by the range of `int64`.

# ✅ Features

- EpochTime a second precise POSIX time wrapper
- LongEpochTime a millisecond precise POSIX time wrapper
- Extension methods for build in time handling classes:
  - DateTime
  - TimeSpan
  - DateOnly
  - TimeOnly
- Fully tested ;)

# 🏃‍♂️ Quick start

## Add the package to yor project

The preferred way of integrating the library is to use the [nuget package](https://www.nuget.org/packages/Epoch.net). This is easily done be either: 

```shell
dotnet add package Epoch.net
```

or

```shell
Install-Package Epoch.net
```

## Reference

After the package is added to the project add the using directive:

```csharp
...
using Epoch.net;
...
```

Done 🎉

# 🦮 Helping out

Everyone can help in their own way. 

Here are just some ideas:

- Create a (ticket)[https://github.com/dejanfajfar/epoch.net/issues] with a improvement suggestion
- Create a (ticket)[https://github.com/dejanfajfar/epoch.net/issues] with a bug or issue you may be having
- Create a (ticket)[https://github.com/dejanfajfar/epoch.net/issues] with a question that you may be having
- Have a look at the code and suggest improvements
- Spread the word of its existence
- And for the hardcore among you you can help me with the documentation 😈

# ⏭️ Next steps

- Consult the [Wiki](https://github.com/dejanfajfar/epoch.net/wiki)
- Create a (ticket)[https://github.com/dejanfajfar/epoch.net/issues] to help the project improve

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.0]
### Changed

- Updated to .net version 7
- Moved build to github actions
- Updated code formatting to new standard

### Added

- DateOnly extensions
- TimeOnly extensions

### Removed

- The Constants static class

## [3.1]
### Added

- LongEpochTime ToString method
- EpochTime ToString method

## [3.0]
### Added

- LongEpochTime
- EpochTime.ShortEpoch
- EpochTime.RawEpoch
- EpochTime.DateTime
- EpochTime.TimeSpan
- EpochTime.Add
- EpochValidator
- EpochTimeValueException
- ITimeProvider
- DefaultTimeProvider
- Long extensions methods

### Removed

- EpochOverflowException
- EpochTime.AddSeconds
- EpochTime.ToRawEpoch
- EpochTime.ToDateTime
- EpochTime.ToTimeSpan

## [2.0]
### Changed

- Renamed the ```Epoch``` class to ```EpochTime``` to fix the name collision with the namespace.
- Updated the README.md to reflect the new naming

### Fixed

- The name collision bug

## [1.1]
### Fixed

* DateTime.ToRawEpoch now thrown an EpochOverflow exception if the resulting epoch is not storable in an integer


Epoch.net/EpochTime.cs
using System;

namespace Epoch.net;

/// <summary>
/// Implements a to the second precise unix timestamp
/// </summary>
public sealed class EpochTime
{
    static EpochTime() => TimeProvider = new DefaultTimeProvider();

    /// <summary>
    /// Provides a new instance of a <see cref="EpochTime"/> with the maximal possible value
    /// </summary>
    public static EpochTime MAX => new(MAX_VALUE);
    /// <summary>
    /// Provides a new instance of a <see cref="EpochTime"/> with the minimal possible value
    /// </summary>
    public static EpochTime MIN => new(MIN_VALUE);
    /// <summary>
    /// The minimal raw epoch value
    /// </summary>
    public const int MIN_VALUE = -2147483648;
    /// <summary>
    /// The maximal raw epoch value
    /// </summary>
    public const int MAX_VALUE = 2147483647;
    /// <summary>
    /// The maximal <see cref="DateTime"/> value that a <see cref="EpochTime"/> can take
    /// </summary>
    public static readonly DateTime MAX_DATETIME = new(2038, 1, 19,3,14,07, DateTimeKind.Utc); // 3:14:07 Tuesday, 19 January 2038 UTC
    /// <summary>
    /// The minimal <see cref="DateTime"/> value that a <see cref="EpochTime"/> can take
    /// </summary>
    public static readonly DateTime MIN_DATETIME = new(1901, 12, 13, 20, 45, 52, DateTimeKind.Utc); // 20:45:52 Friday, 13 December 1901 UTC

    /// <summary>
    /// The default value of <see cref="LongEpochTime"/>
    /// </summary>
    /// <remarks>
    /// Thursday, January 1, 1970 12:00:00 AM GMT
    /// </remarks>
    public static EpochTime Default => new(0);

    private static IDateTimeProvider TimeProvider { get; set; }

    #region Constructors

    /// <summary>
    /// Creates a new instance of <see cref="EpochTime"/> with a given rawEpoch
    /// </summary>
    /// <param name="rawEpoch">The number of seconds from 1970-01-01T00:00:00</param>
    public EpochTime(int rawEpoch) => this.Epoch = rawEpoch;

    /// <summary>
    /// Creates a new instance of <see cref="EpochTime"/> with the given <see cref="DateTime"/>
    /// </summary>
    /// <param name="dateTime">The <see cref="DateTime"/> used to initialize the <see cref="EpochTime"/></param>
    /// <exception cref="ArgumentNullException">
    /// If the passed <see cref="DateTime"/> is outside of the valid Epoch range
    /// </exception>
    /// <remarks>
    ///    The given <see cref="DateTime"/> is converted to UTC according to the environment settings
    /// </remarks>
    public EpochTime(DateTime dateTime)
    {
        if (!dateTime.IsValidEpochTime())
        {
            throw new EpochTimeValueException(dateTime);
        }

        Epoch = dateTime.ToEpochTimestamp();
    }

    /// <summary>
    /// Creates a new instance of <see cref="EpochTime"/> with the given <see cref="EpochTime"/> instance as initialisation data
    /// </summary>
    /// <param name="epoch"><see cref="EpochTime"/> used to initialize the instance</param>
    /// <exception cref="ArgumentNullException">If the epoch instance is null</exception>
    /// <remarks>
    /// Is a simple copy constructor
    /// </remarks>
    public EpochTime(EpochTime epoch)
    {
        if (epoch == null)
        {
            throw new ArgumentNullException(nameof(epoch), "The provided epoch can not be null");
        }

        Epoch = epoch.Epoch;
    }

    /// <summary>
    /// Creates a new instance of <see cref="EpochTime"/> with the given <see cref="TimeSpan"/> as a offset from 1970-01-01T00:00Z
    /// </summary>
    /// <param name="timeSpan">The <see cref="TimeSpan"/> representing the offset from 1970-01-01T00:00Z</param>
    /// <exception cref="EpochTimeValueException">
    /// If the offset would put the <see cref="EpochTime"/> outside of the valid Epoch range
    /// </exception>
    public EpochTime(TimeSpan timeSpan)
    {
        if (!timeSpan.IsValidEpochTime())
        {
            throw new EpochTimeValueException(timeSpan);
        }

        Epoch = timeSpan.ToEpochTimestamp();
    }
    #endregion

    #region Static methods

    /// <summary>
    /// Injects a new global thread safe <see cref="TimeProvider"/> instance to be used globally
    /// </summary>
    /// <param name="timeProvider">The new time provider instance</param>
    public static void SetTimeProvider(IDateTimeProvider timeProvider)
    {
        if (timeProvider != null)
        {
            TimeProvider = timeProvider;
        }
    }

    /// <summary>
    /// Resets the global time provider to the default system time provider 
    /// </summary>
    public static void ResetTimeProvider() => TimeProvider = new DefaultTimeProvider();

    /// <summary>
    /// Gets the current UTC date and time as an <see cref="EpochTime"/>
    /// </summary>
    public static EpochTime Now => TimeProvider.UtcNow.ToEpochTime();

    #endregion

    /// <summary>
    /// Gets a <see cref="int"/> representation of the <see cref="EpochTime"/> instance
    /// </summary>
    public int Epoch { get; private set; }

    /// <summary>
    /// Returns a <see cref="DateTime"/> representation of the <see cref="EpochTime"/> instance
    /// </summary>
    public DateTime DateTime => Epoch.ToDateTime();

    /// <summary>
    /// Returns a <see cref="TimeSpan"/> representation of the <see cref="EpochTime"/> instance
    /// </summary>
    public TimeSpan TimeSpan => Epoch.ToTimeSpan();

#if NET6_0_OR_GREATER
    /// <summary>
    /// Returns a <see cref="DateOnly"/> representation of the <see cref="EpochTime"/> instance
    /// </summary>
    public DateOnly DateOnly => DateOnly.FromDateTime(Epoch.ToDateTime());

    /// <summary>
    /// Returns a <see cref="TimeOnly"/> representation of the <see cref="EpochTime"/> instance
    /// </summary>
    public TimeOnly TimeOnly => TimeOnly.FromDateTime(Epoch.ToDateTime());

#endif

    #region Epoch manipulation

    /// <summary>
    /// Applies the given <see cref="TimeSpan"/> offset to the <see cref="EpochTime"/> instance
    /// </summary>
    /// <param name="span">The <see cref="TimeSpan"/> offset to apply</param>
    /// <returns>The updated <see cref="EpochTime"/> instance</returns>
    /// <remarks>
    ///  The method returns an updated <see cref="EpochTime"/> but also updates the calling instance
    /// </remarks>
    public EpochTime Add(TimeSpan span)
    {
        var newSpan = TimeSpan + span;
        Epoch = newSpan.ToEpochTimestamp();
        return this;
    }

    #endregion

    #region Operators

    /// <summary>
    /// Implements the + operator between two <see cref="EpochTime"/> instances
    /// </summary>
    /// <param name="operand1">The augend</param>
    /// <param name="operand2">The addend</param>
    /// <returns>The sum of the two <see cref="EpochTime"/></returns>
    /// <exception cref="EpochTimeValueException">
    /// If the result is not a valid <see cref="EpochTime"/>
    /// </exception>
    public static EpochTime operator +(EpochTime operand1, EpochTime operand2)
    {
        long epochSum = operand1.Epoch + (long)operand2.Epoch;

        return epochSum.IsValidEpochTimestamp() is false
            ? throw new EpochTimeValueException(epochSum)
            : new EpochTime(Convert.ToInt32(epochSum));
    }

    /// <summary>
    /// Implements the - operator between two <see cref="EpochTime"/> instances
    /// </summary>
    /// <param name="operand1">The minuend</param>
    /// <param name="operand2">The subtrahend</param>
    /// <returns>The Difference between the two <see cref="EpochTime"/> instances</returns>
    /// <exception cref="EpochTimeValueException">
    /// If the result is not a valid <see cref="EpochTime"/></exception>
    public static EpochTime operator -(EpochTime operand1, EpochTime operand2)
    {
        long epochSub = operand1.Epoch - (long)operand2.Epoch;

        return epochSub.IsValidEpochTimestamp() is false
            ? throw new EpochTimeValueException(epochSub)
            : new EpochTime(operand1.Epoch - operand2.Epoch);
    }
    #endregion

    #region Equals

    /// <inheritdoc/>
    public override bool Equals(object obj) => obj is EpochTime comparedEpoch && comparedEpoch.Epoch == Epoch;

    /// <inheritdoc/>
    public override int GetHashCode() => Epoch;

    /// <inheritdoc/>
    public override string ToString() => Epoch.ToString();

    #endregion
}